src/compiler/tsbuildPublic.ts

```diff
@@ -1834,6 +1835,22 @@ function getUpToDateStatusWorker<T extends BuilderProgram>(state: SolutionBuilde
             newestInputFileName = inputFile;
             newestInputFileTime = inputTime;
         }
+
+        if (buildInfoProgram) seenRoots.add(toPath(state, inputFile));
+    }
+
+    if (buildInfoProgram) {
+        if (!buildInfoVersionMap) buildInfoVersionMap = getBuildInfoFileVersionMap(buildInfoProgram, buildInfoPath!, host);
+        for (const existingRoot of buildInfoVersionMap.roots) {
+            if (!seenRoots.has(existingRoot)) {

```

Comment:
We don't do this for the opposite direction, i.e. if the old program had some file that the new one does not?

Reply:
We are checking if new program doesnt have a root that old program has
The check for new program having extra root happens when checking inputFile -> if it was already part of the old program here: https://github.com/microsoft/TypeScript/pull/52417/files#diff-0558af289c76f66611354da63a14152fe5cd3a72b3c3e59eb880808efc2df442L1824